src/util/bitset.h

```diff
@@ -0,0 +1,463 @@
+        {
+            return a.m_idx == N;
+        }
+        /** Progress to the next 1 bit (only if != IteratorEnd). */
+        constexpr Iterator& operator++() noexcept
+        {
+            m_val &= m_val - I{1U};
+            if (m_val == 0) {
+                while (true) {
+                    ++m_idx;
+                    if (m_idx == N) break;
+                    m_val = (*m_ptr)[m_idx];
+                    if (m_val) {
+                        m_pos = std::countr_zero(m_val) + m_idx * LIMB_BITS;
+                        break;
+                    }
+                }
+            } else {
+                m_pos = std::countr_zero(m_val) + m_idx * LIMB_BITS;
+            }
+            return *this;
+        }
+        /** Get the current bit position (only if != IteratorEnd). */
+        constexpr const unsigned& operator*() const noexcept { return m_pos; }
+    };
+
+public:
+    /** Construct an all-zero bitset. */
+    constexpr MultiIntBitSet() noexcept : m_val{} {}
+    /** Copy construct a bitset. */
+    constexpr MultiIntBitSet(const MultiIntBitSet&) noexcept = default;
+    /** Copy assign a bitset. */
+    constexpr MultiIntBitSet& operator=(const MultiIntBitSet&) noexcept = default;
+    /** Set a bit to 1. */
+    void constexpr Set(unsigned pos) noexcept { m_val[pos / LIMB_BITS] |= I{1U} << (pos % LIMB_BITS); }
+    /** Set a bit to the specified value. */
+    void constexpr Set(unsigned pos, bool val) noexcept { m_val[pos / LIMB_BITS] = (m_val[pos / LIMB_BITS] & ~I(I{1U} << (pos % LIMB_BITS))) | (I{val} << (pos % LIMB_BITS)); }
+    /** Construct a bitset from a list of values. */
+    constexpr MultiIntBitSet(std::initializer_list<unsigned> ilist) noexcept : m_val{}
+    {
+        for (auto i : ilist) Set(i);
+    }
+    /** Set a bitset to a list of values. */
+    constexpr MultiIntBitSet& operator=(std::initializer_list<unsigned> ilist) noexcept
+    {
+        m_val.fill(0);
+        for (auto i : ilist) Set(i);
+        return *this;
+    }
+    /** Set a bit to 0. */
+    void constexpr Reset(unsigned pos) noexcept { m_val[pos / LIMB_BITS] &= ~I(I{1U} << (pos % LIMB_BITS)); }
+    /** Retrieve a bit at the given position. */
+    bool constexpr operator[](unsigned pos) const noexcept { return (m_val[pos / LIMB_BITS] >> (pos % LIMB_BITS)) & 1U; }
+    /** Construct a bitset with the singleton i. */
+    static constexpr MultiIntBitSet Singleton(unsigned i) noexcept
+    {
+        MultiIntBitSet ret;
+        ret.m_val[i / LIMB_BITS] = I{1U} << (i % LIMB_BITS);
+        return ret;
+    }
+    /** Construct a bitset with bits 0..count-1 (inclusive) set to 1. */
+    static constexpr MultiIntBitSet Fill(unsigned count) noexcept
+    {
+        MultiIntBitSet ret;
+        if (count) {
+            unsigned i = 0;
+            while (count > LIMB_BITS) {
+                ret.m_val[i++] = ~I{0};
+                count -= LIMB_BITS;
+            }
+            ret.m_val[i] = I(~I{0}) >> (LIMB_BITS - count);
```

Comment:
Added `std::conditional_t<(BITS <= 16), bitset_detail::IntBitSet<uint16_t>, ... ` and removed the `I()`, wrote unit test in bc36241f5a0a23d16645155eb819e94b0db2813e, and it indeed proves you are correct. Right now it cannot occur since `IntBitSet<uint32_t>` is the smallest we go, but better not to leave loaded footguns lying around.

<details>
<summary>Test Log</summary>

```
$ src/test/test_bitcoin -t bitset_tests
...
Running 1 test case...
test/bitset_tests.cpp(18): error: in "bitset_tests/bitset_test_15": check std::popcount(sim) == set.Count() has failed [1 != 16]
test/bitset_tests.cpp(18): error: in "bitset_tests/bitset_test_15": check std::popcount(sim) == set.Count() has failed [2 != 16]
test/bitset_tests.cpp(18): error: in "bitset_tests/bitset_test_15": check std::popcount(sim) == set.Count() has failed [3 != 16]
test/bitset_tests.cpp(18): error: in "bitset_tests/bitset_test_15": check std::popcount(sim) == set.Count() has failed [4 != 16]
test/bitset_tests.cpp(18): error: in "bitset_tests/bitset_test_15": check std::popcount(sim) == set.Count() has failed [5 != 16]
test/bitset_tests.cpp(18): error: in "bitset_tests/bitset_test_15": check std::popcount(sim) == set.Count() has failed [6 != 16]
test/bitset_tests.cpp(18): error: in "bitset_tests/bitset_test_15": check std::popcount(sim) == set.Count() has failed [7 != 16]
test/bitset_tests.cpp(18): error: in "bitset_tests/bitset_test_15": check std::popcount(sim) == set.Count() has failed [8 != 16]
test/bitset_tests.cpp(18): error: in "bitset_tests/bitset_test_15": check std::popcount(sim) == set.Count() has failed [9 != 16]
test/bitset_tests.cpp(18): error: in "bitset_tests/bitset_test_15": check std::popcount(sim) == set.Count() has failed [10 != 16]
test/bitset_tests.cpp(18): error: in "bitset_tests/bitset_test_15": check std::popcount(sim) == set.Count() has failed [11 != 16]
test/bitset_tests.cpp(18): error: in "bitset_tests/bitset_test_15": check std::popcount(sim) == set.Count() has failed [12 != 16]
test/bitset_tests.cpp(18): error: in "bitset_tests/bitset_test_15": check std::popcount(sim) == set.Count() has failed [13 != 16]
test/bitset_tests.cpp(18): error: in "bitset_tests/bitset_test_15": check std::popcount(sim) == set.Count() has failed [14 != 16]
test/bitset_tests.cpp(18): error: in "bitset_tests/bitset_test_15": check std::popcount(sim) == set.Count() has failed [15 != 16]
```
</details>